File: src/main/java/com/naminhyeok/fantazzk/room/domain/AuctionRound.java
package com.naminhyeok.fantazzk.room.domain;

import org.springframework.lang.Nullable;

public final class AuctionRound {
    private final int round;
    private final RoomBid highestBid;

    public AuctionRound(int round, RoomBid highestBid) {
        require(round > 0, "경매 라운드는 1 이상이어야 합니다");
        this.round = round;
        this.highestBid = highestBid;
    }

    public int getRound() {
        return round;
    }

    @Nullable
    public RoomBid getHighestBid() {
        return highestBid;
    }

    public void requireHigherBid(int amount) {
        require(amount > highestAmount(), "현재 최고가보다 높아야 합니다");
    }

    public void requireRosterCapacity(int currentMemberCount, int picksPerTeam) {
        checkState(currentMemberCount < picksPerTeam, "팀장의 팀원 정원이 가득 찼습니다");
    }

    public AuctionRoundSettlement settle(String playerName, int assignedCountAfterSettlement, int totalRequired) {
        int nextRound = round + 1;
        if (highestBid != null) {
            return new AuctionRoundSettlement(
                    playerName,
                    AuctionOutcome.SOLD,
                    nextRound,
                    assignedCountAfterSettlement >= totalRequired,
                    highestBid);
        }
        return new AuctionRoundSettlement(
                playerName,
                AuctionOutcome.PASSED,
                nextRound,
                false,
                null);
    }

    private int highestAmount() {
        return highestBid == null ? 0 : highestBid.getAmount();
    }

    private static void require(boolean condition, String message) {
        if (!condition) {
            throw new IllegalArgumentException(message);
        }
    }

    private static void checkState(boolean condition, String message) {
        if (!condition) {
            throw new IllegalStateException(message);
        }
    }
}

File: src/main/kotlin/com/naminhyeok/fantazzk/room/AuctionRound.kt
package com.naminhyeok.fantazzk.room

data class AuctionRound(
    val round: Int,
    val highestBid: RoomBid? = null,
) {
    init {
        require(round > 0) { "경매 라운드는 1 이상이어야 합니다" }
    }

    fun requireHigherBid(amount: Int) {
        require(amount > highestAmount()) { "현재 최고가보다 높아야 합니다" }
    }

    fun requireRosterCapacity(
        currentMemberCount: Int,
        picksPerTeam: Int,
    ) {
        check(currentMemberCount < picksPerTeam) { "팀장의 팀원 정원이 가득 찼습니다" }
    }

    fun settle(
        playerName: String,
        assignedCountAfterSettlement: Int,
        totalRequired: Int,
    ): AuctionRoundSettlement {
        val nextRound = round + 1
        return if (highestBid != null) {
            AuctionRoundSettlement(
                playerName = playerName,
                outcome = AuctionOutcome.SOLD,
                nextRound = nextRound,
                completed = assignedCountAfterSettlement >= totalRequired,
                winningBid = highestBid,
            )
        } else {
            AuctionRoundSettlement(
                playerName = playerName,
                outcome = AuctionOutcome.PASSED,
                nextRound = nextRound,
                completed = false,
                winningBid = null,
            )
        }
    }

    private fun highestAmount(): Int = highestBid?.amount ?: 0
}

data class AuctionRoundSettlement(
    val playerName: String,
    val outcome: AuctionOutcome,
    val nextRound: Int,
    val completed: Boolean,
    val winningBid: RoomBid?,
)
